[addons][tools] add string utils

AlwinEsch opened the pull request:
## Description
As the p8-platform brings problems becomes here a own StringUtils class added to addon headers.
This is mostly based upon original of Kodi.

Related new docs are visible here http://alwinesch.github.io/group__cpp__kodi__tools___string_utils.html

Commit 1:
-------------------------
**[addons][tools] add own StringUtils.h to help addons about**

This is the part most often used by p8-platform to edit strings in C++.
With this, it is added to the DevKit with.

It's based 99% on that of Kodi and is more or less a copy of it.
Unfortunately Kodi's own cannot be used in addons because it then breaks the "C" ABI.

It works completely independently of Kodi, so there are a few fewer functions in it than in the original (but these were not on p8-platform either).

The background is the confusion that currently exists with p8-platform. This is based on Kodi, but with its last release (October 26th, 2016) it is very old and brings problems with some OS.
This in turn resulted in a clone from Kodi which contains fixes.
Now p8-platform will be continued again (without a release) which in turn does not correspond to our necessary changes and does not make direct sense on the external side (e.g. TVOS).
Linux distributions continue to use the release of it and we need an extension to it for all OS.

Therefore it is brought into Kodi's addon header so that it is always up to date.

In addition, other format functions (e.g. `kodi::Log()`) in the dev kit can then also be used by the new system.


Commit 2:
-------------------------
**[addons] use new StringUtils::Format within addon header functions**

This add the use of new StringUtils.h for format to `kodi::Log` and `kodi::QueueFormattedNotification`. After them is it more safe as no fixed memories are used.

Commit 3:
-------------------------
**[addons][tools] add initial new documentations to StringUtils**

This add a big new documentation to the string utils functions.
Some parts are take from source by Kodi and some are complete new.

About my bad english need it be improved in future and maybe to correct also wrong descriptions.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons] use new StringUtils::Format within addon header functions

This add the use of new StringUtils.h for format to kodi::Log and
kodi::QueueFormattedNotification. After them is it more safe as no fixed memories
are used.

wsnipex commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/StringUtils.h:
```diff
@@ -0,0 +1,3098 @@
+#include <string>
+#include <vector>
+
+#include <fstrcmp.h>
```
this is libfstrcmp, is this really needed?

AlwinEsch replied:
Have removed it, was gone down that "fstrcmp.h" use a own lib.
Maybe if needed in future it can be added as "C" interface function Kodi (or the addon it himself :smile:)

phunkyfish reviewed (approved):
Looks good to me

AlwinEsch commented:
jenkins build this please

wsnipex commented on xbmc/addons/kodi-dev-kit/include/kodi/tools/StringUtils.h:
```diff
@@ -0,0 +1,3014 @@
+#endif // PRINTF_FORMAT_STRING
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(X) (sizeof(X) / sizeof((X)[0]))
```
this causes a macro redefinition warning: https://jenkins.kodi.tv/job/Android-ARM/19462/clang/new/source.317b30d6-037d-474c-b6e3-d4dfabb48eb0/#20

AlwinEsch replied:
Was by functions used where removed yesterday, this is also kicked out now.

Thanks a lot to remark it!

AlwinEsch commented:
After them is in comes another for thread, see https://github.com/AlwinEsch/kodi/blob/add-thread-to-dev-kit/xbmc/addons/kodi-dev-kit/include/kodi/tools/Thread.h

Seen by change of pvr.argustv that it can better to have as class, as there was not easily to do with std::thread alone. Also helpfull for other addons where use threads.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this with addons please

wsnipex commented:
heh, looks like peripheral.joystick has it's own StringUtils ever since ;) https://github.com/xbmc/peripheral.joystick/blob/Matrix/src/utils/StringUtils.cpp

AlwinEsch commented:
Wow :smile:, yes and three other functions where not here.
I take it from it and add also to here.

ping @garbear I steal your code :wink:

AlwinEsch commented:
On peripheral.joystick is also this C++17 error in his StringUtils :sweat_smile: 

```
/Users/Shared/jenkins/workspace/OSX-64/tools/depends/target/binary-addons/macosx10.14_x86_64-target-debug/build/peripheral.joystick/src/utils/StringUtils.cpp:86:78: error: no member named 'ptr_fun' in namespace 'std'
  str.erase(str.begin(), std::find_if(str.begin(), str.end(), std::not1(std::ptr_fun(isspace_c))));
                                                                        ~~~~~^
/Users/Shared/jenkins/workspace/OSX-64/tools/depends/target/binary-addons/macosx10.14_x86_64-target-debug/build/peripheral.joystick/src/utils/StringUtils.cpp:92:67: error: no member named 'ptr_fun' in namespace 'std'
  str.erase(std::find_if(str.rbegin(), str.rend(), std::not1(std::ptr_fun(isspace_c))).base(), str.end());
                                                             ~~~~~^
2 errors generated.
```

AlwinEsch committed:
[addons][tools] add own StringUtils.h to help addons about

This is the part most often used by p8-platform to edit strings in C++.
With this, it is added to the DevKit with.

It's based 99% on that of Kodi and is more or less a copy of it.
Unfortunately Kodi's own cannot be used in addons because it then breaks the "C" ABI.

It works completely independently of Kodi, so there are a few fewer functions in it than in the original (but these were not on p8-platform either).

The background is the confusion that currently exists with p8-platform. This is based on Kodi,
but with its last release (October 26th, 2016) it is very old and brings problems with some OS.
This in turn resulted in a clone from Kodi which contains fixes.
Now p8-platform will be continued again (without a release) which in turn does not correspond
to our necessary changes and does not make direct sense on the external side (e.g. TVOS).
Linux distributions continue to use the release of it and we need an extension to it for all OS.

Therefore it is brought into Kodi's addon header so that it is always up to date.

In addition, other format functions (e.g. `kodi::Log()`) in the dev kit can then also be used by the new system.

AlwinEsch commented:
jenkins build this please

garbear reviewed (approved):
> On peripheral.joystick is also this C++17 error in his StringUtils sweat_smile
> 
> ```
> /Users/Shared/jenkins/workspace/OSX-64/tools/depends/target/binary-addons/macosx10.14_x86_64-target-debug/build/peripheral.joystick/src/utils/StringUtils.cpp:86:78: error: no member named 'ptr_fun' in namespace 'std'
>   str.erase(str.begin(), std::find_if(str.begin(), str.end(), std::not1(std::ptr_fun(isspace_c))));
>                                                                         ~~~~~^
> /Users/Shared/jenkins/workspace/OSX-64/tools/depends/target/binary-addons/macosx10.14_x86_64-target-debug/build/peripheral.joystick/src/utils/StringUtils.cpp:92:67: error: no member named 'ptr_fun' in namespace 'std'
>   str.erase(std::find_if(str.rbegin(), str.rend(), std::not1(std::ptr_fun(isspace_c))).base(), str.end());
>                                                              ~~~~~^
> 2 errors generated.
> ```

It's all over the place!!!

Nice work

AlwinEsch committed:
[addons][tools] add initial new documentations to StringUtils

This add a big new documentation to the string utils functions.
Some parts are take from source by Kodi and some are complete new.

About my bad english need it be improved in future and maybe to correct
also wrong descriptions.

AlwinEsch commented:
jenkins build this please
